src/common/meta/src/key/tombstone.rs

```diff
@@ -0,0 +1,518 @@
+    ) {
+        let key = self.get_inner().clone();
+        (TxnOp::Get(key.clone()), TxnOpGetResponseSet::filter(key))
+    }
+}
+
+fn format_on_failure_error_message<F: FnMut(&mut TxnOpGetResponseSet) -> Option<Vec<u8>>>(
+    mut set: TxnOpGetResponseSet,
+    on_failure_kv_and_filters: Vec<(Vec<u8>, Vec<u8>, F)>,
+) -> String {
+    on_failure_kv_and_filters
+        .into_iter()
+        .flat_map(|(key, value, mut filter)| {
+            let got = filter(&mut set);
+            let Some(got) = got else {
+                return Some(format!(
+                    "For key: {} was expected: {}, but value does not exists",
+                    String::from_utf8_lossy(&key),
+                    String::from_utf8_lossy(&value),
+                ));
+            };
+
+            if got != value {
+                Some(format!(
+                    "For key: {} was expected: {}, but got: {}",
+                    String::from_utf8_lossy(&key),
+                    String::from_utf8_lossy(&value),
+                    String::from_utf8_lossy(&got),
+                ))
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>()
+        .join("; ")
+}
+
+impl TombstoneManager {
+    /// Returns [TombstoneManager].
+    pub fn new(kv_backend: KvBackendRef) -> Self {
+        Self { kv_backend }
+    }
+
+    /// Creates tombstones for keys.
+    ///
+    /// Preforms to:
+    /// - retrieve all values corresponding `keys`.
+    /// - stores tombstone values.
+    pub(crate) async fn create(&self, keys: Vec<Key>) -> Result<()> {
+        // Builds transaction to retrieve all values
+        let (operations, mut filters): (Vec<_>, Vec<_>) =
+            keys.iter().map(|key| key.build_get_op()).unzip();
+
+        let txn = Txn::new().and_then(operations);
+        let mut resp = self.kv_backend.txn(txn).await?;
+        ensure!(
+            resp.succeeded,
+            error::UnexpectedSnafu {
+                err_msg: "Failed to retrieves the metadata"
+            }
+        );
+
+        let mut set = TxnOpGetResponseSet::from(&mut resp.responses);
+        // Builds the create tombstone transaction.
+        let mut tombstone_operations = Vec::with_capacity(keys.len() * 2);
+        let mut tombstone_comparison = vec![];
+        let mut on_failure_operations = vec![];
+        let mut on_failure_kv_and_filters = vec![];
+        for (idx, key) in keys.iter().enumerate() {
+            let filter = &mut filters[idx];
+            let value = filter(&mut set).with_context(|| error::UnexpectedSnafu {
+                err_msg: format!(
+                    "Missing value, key: {}",
+                    String::from_utf8_lossy(key.get_inner())
+                ),
+            })?;
+            let (origin_key, tombstone_key) = TombstoneKey(key.get_inner()).into_keys();
+            // Compares the atomic key.
+            if key.is_atomic() {
+                tombstone_comparison.push(Compare::with_value(
```

Comment:
Besides checking the value's equivalent to the original key, is it necessary to check the absence of the key? Otherwise it might be set by other txn.